### PR TITLE
FFT CONVOLUTION BACKEND FOR 2D FILTERS

### PR DESCRIPTION
**Pull request by codeguitarmaster**
---

This PR introduces FFT BASED convolution for filter2D. For now it is specifically a separate function and only for 2D FILTERS. Once merged, I would make a function for 3d filters also.
I have provided unit tests for this function

Fixes https://github.com/kornia/kornia/issues/1731
---

## 🧪 How Was This Tested?
- [.] **Unit Tests:** (List new/updated tests)
- [.] **Manual Verification:** (Describe the steps you took)
- [.] **Performance/Edge Cases:** (How does this handle nulls, large data, etc.?)

---

## 🕵️ AI Usage Disclosure
*Check one of the following:*
- [ ] 🟢 **No AI used.**
- [.] 🟡 **AI-assisted:** I used AI for boilerplate/refactoring but have manually reviewed and tested every line.
- [ ] 🔴 **AI-generated:** (Note: These PRs may be subject to stricter scrutiny or immediate closure if the logic is not explained).

---

---

### CODE REVIEW

**Comment by codeguitarmaster** [comment]
> Nice addition overall... FFT backend + tests are a good direction ig A few blockers before this can land:
> 
> 1. PR bot: please add Fixes #1731 in the PR description (the validator only catches that exact format).
> 2. filter2d groups got broken: looks like there are two conv2d calls and the second one drops groups=..., which changes semantics (mixes channels). I think the extra call should be removed and groups kept
> 3. dtype: fft_conv forces .float() will change results for float64 (and generally doesnt preserve dtype). can we keep dtype when possible or upcast only for fp16/bf16, then cast back?
> 4. export: tests import fft_conv from kornia.filters, but it doesn’t look exported yet needs kornia/filters/**init**.py update
> 5. speed test the wall clock “FFT must be faster” unit test will be flaky in CI better to drop it or mark as benchmark/skip
>    (minor) behaviour validation error message says padding mode should say “behaviour”....
>    Once the above is fixed... LGTM from my side

Hey I believe I have made the fixes

File: kornia/filters/filter.py [hunk 1]
```diff
@@ -305,3 +306,144 @@ def filter3d(
...
+        str(border_type).lower() in _VALID_BORDERS,
+        f"Invalid border, gotcha {border_type}. Expected one of {_VALID_BORDERS}",
+    )
+    KORNIA_CHECK(
+        str(padding).lower() in _VALID_PADDING,
+        f"Invalid padding mode, gotcha {padding}. Expected one of {_VALID_PADDING}",
+    )
+    KORNIA_CHECK(
+        str(behaviour).lower() in _VALID_BEHAVIOUR,
+        f"Invalid padding mode, gotcha {behaviour}. Expected one of {_VALID_BEHAVIOUR}",
```
Comment (Meggison) [inline kornia/filters/filter.py]: why use gotcha here?

File: kornia/filters/__init__.py [hunk 2]
```diff
@@ -34,7 +34,7 @@
 )
 from .canny import Canny, canny
 from .dissolving import StableDiffusionDissolving
-from .filter import filter2d, filter2d_separable, filter3d
+from .filter import fft_conv, filter2d, filter2d_separable, filter3d
```
Comment (Meggison) [inline kornia/filters/__init__.py:37]: Need to add __init__ file to __all__

File: kornia/filters/filter.py [hunk 3]
```diff
@@ -305,3 +306,144 @@ def filter3d(
...
+    )
+    # prepare kernel
+    n = input.ndim - 2
+    stride_ = n * (1,)
+    b, c, h, w = input.shape
+    if str(behaviour).lower() == "conv":
+        tmp_kernel = kernel.flip((-2, -1))[:, None, ...].to(device=input.device, dtype=input.dtype)
+    else:
+        tmp_kernel = kernel[:, None, ...].to(device=input.device, dtype=input.dtype)
+        #  str(behaviour).lower() == 'conv':
```
Comment (Meggison) [inline kornia/filters/filter.py]: remove

File: kornia/filters/filter.py [hunk 4]
```diff
@@ -305,3 +306,144 @@ def filter3d(
...
+    """Multiplies two complex-valued tensors."""
+    a = a.view(a.size(0), groups, -1, *a.shape[2:])
+    b = b.view(groups, -1, *b.shape[1:])
+    a = torch.movedim(a, 2, a.dim() - 1).unsqueeze(-2)
+    b = torch.movedim(b, (1, 2), (b.dim() - 1, b.dim() - 2))
+    real = a.real @ b.real - a.imag @ b.imag
+    imag = a.imag @ b.real + a.real @ b.imag
+    real = torch.movedim(real, real.dim() - 1, 2).squeeze(-1)
+    imag = torch.movedim(imag, imag.dim() - 1, 2).squeeze(-1)
+    c = torch.zeros(real.shape, dtype=torch.complex128, device=a.device)
```
Comment (Meggison) [inline kornia/filters/filter.py]: Might be better to define a complex tensor rather than creating a tensor of zeros and reassigning

File: kornia/filters/filter.py (hunk 1 above)
Comment (manas1245agrawal) [reply kornia/filters/filter.py]: That was something used in the original function of filter 2d but ok I will remove it

File: kornia/filters/filter.py (hunk 4 above)
Comment (manas1245agrawal) [reply kornia/filters/filter.py]: Okay okay, Other than this is everything alright sir

Comment (Meggison) [reply kornia/filters/filter.py]: Oh should be good

File: kornia/filters/filter.py (hunk 1 above)
Comment (Meggison) [reply kornia/filters/filter.py]: I see, might be good to change it still. Seems a bit vague

File: kornia/filters/filter.py (hunk 4 above)
Comment (codeguitarmaster) [reply kornia/filters/filter.py]: hey @Meggison I believe I have made the changes, please check

**Comment by codeguitarmaster** [comment]
@edgarriba
Thanks for the review 
Actually somehow I was getting errors continously using the torch complex multiplication, than I decided to go with manual. The speed difference is very small I assure you. But I am still working on that.
Okay I will fix the typo.
Actually that was something I was doing earlier to prevent some changing of dtypes, but I shall remove it since the previous errors are fixed.
Yes I have provided the benchmarks on discord - https://colab.research.google.com/drive/136t4ce0b-5GHGh7kzQ3JcGxZRrb7Ge-M?usp=sharing
I have done three tests in this benchmark - There are three tests performed here -
To ensure that on random filters and inputs the results are same ( ran 120 loops)
The averaga time taken for an fft_conv to run and filter2d to run (ran 33 times and took the avg)
A graph between time and kernel size clearly showing the speed of fft_conv
Okay I will do the torch matmul function private
Okay I shall document the limitation.

**Comment by codeguitarmaster** [comment]
@edgarriba
I have made the changes

**Comment by edgarriba** [comment]
@codeguitarmaster CI looks green ✅ — good progress! 

Before this can be merged, please make sure the following are addressed:
1. Use `torch.view_as_complex` / `torch.view_as_real` (or `torch.fft.rfft2`/`torch.fft.irfft2`) rather than manual complex arithmetic — this keeps the code idiomatic and ensures grad compatibility.
2. Double-check the padding behavior matches the existing `filter2d` semantics exactly (same padding, border handling).
3. Add a brief docstring explaining when FFT-based convolution is preferable (large kernels).

@sidd-27 / @cjpurackal — once the contributor addresses the above, can you do a final pass?

**Comment by codeguitarmaster** [comment]
@sidd-27
 @cjpurackal
 @Meggison 
@edgarriba
I believe the changes are done.
I have added the doc string
I have double checked the padding mode
I have removed the complex arithmetic to torch way of a @ b. The reason for using movedim is to convert into groups before multiplication.

File: kornia/filters/filter.py [hunk 5]
```diff
@@ -305,3 +305,146 @@ def filter3d(
...
+
+def _complex_matmul(a: Tensor, b: Tensor, groups: int = 1) -> Tensor:
+    """Multiplies two complex-valued tensors using einsum."""
+    B = a.size(0)
+    a = a.view(B, groups, -1, *a.shape[2:])
+    b = b.view(groups, -1, *b.shape[1:])
+    # b g i ... , g o i ...  -> b g o ...
+    c = torch.einsum("bgix..., goix... -> bgox...", a, b)
+    c = c.reshape(B, -1, *c.shape[3:])
+    return c
```
Comment (sidd-27) [inline kornia/filters/filter.py]: why not use complex64 data type here?

File: kornia/filters/filter.py [hunk 6]
```diff
@@ -305,3 +305,146 @@ def filter3d(
...
+    output_ = output_.to(dtype=dtype)
+    crop_slices = [slice(None), slice(None)] + [
+        slice(0, (input_size[i] - tmp_kernel.size(i) + 1), stride_[i - 2]) for i in range(2, input.ndim)
+    ]
+    output_ = output_[crop_slices].contiguous()
+    if padding == "same":
+        out_ = output_.view(b, c, h, w)
+    else:
+        out_ = output_.view(b, c, h - height + 1, w - width + 1)
+    return out_
```
Comment (sidd-27) [inline kornia/filters/filter.py]:
```
    b, c, h, w = input.shape
    kh, kw = kernel.shape[-2:]

    if str(behaviour).lower() == "conv":
        tmp_kernel = kernel.flip((-2, -1))[:, None, ...].to(device=input.device, dtype=input.dtype)
    else:
        tmp_kernel = kernel[:, None, ...].to(device=input.device, dtype=input.dtype)

    if normalized:
        tmp_kernel = normalize_kernel2d(tmp_kernel)

    if padding == "same":
        padding_shape = _compute_padding([kh, kw])
        input_padded = F.pad(input, padding_shape, mode=border_type)
    else:
        input_padded = input

    padded_h, padded_w = input_padded.shape[-2:]

    input_fr = torch.fft.rfftn(input_padded, dim=(-2, -1))
    kernel_fr = torch.fft.rfftn(tmp_kernel, s=(padded_h, padded_w), dim=(-2, -1))

    output_fr = input_fr * torch.conj(kernel_fr)
    output = torch.fft.irfftn(output_fr, s=(padded_h, padded_w), dim=(-2, -1))

    crop_h = padded_h - kh + 1
    crop_w = padded_w - kw + 1
    
    return output[..., :crop_h, :crop_w].contiguous()

```

something like this could be faster, could you benchmark the two and implement the faster one?

**Review by sidd-27 — CHANGES_REQUESTED** [review]

File: kornia/filters/filter.py (hunk 5 above)
Comment (manas1245agrawal) [reply kornia/filters/filter.py]: Because it would be forcing a dtype, It the dtype of c is complex128 then it is a downgrade.

Comment (manas1245agrawal) [reply kornia/filters/filter.py]: Or do you mean to start with a and b as torch.complex64 tensors

Comment (codeguitarmaster) [reply kornia/filters/filter.py]: We avoid forcing complex64 to preserve the input dtype. The function respects PyTorch's type promotion rules, so if the inputs are complex128, the output will also remain complex128. Forcing complex64 would implicitly downcast and reduce precision.

**Comment by codeguitarmaster** [comment]
@sidd-27 This was the benchmark result that I got 
image

Your and mine implementation have similar speed. Which one do you want me to keep
(This is on GPU).

**Comment by sidd-27** [comment]
my version seems simpler and easier to maintain down the line

**Comment by codeguitarmaster** [comment]
> my version seems simpler and easier to maintain down the line

Okay no problem, I have updated with your version of the code. If possible plz see

**Comment by codeguitarmaster** [comment]
> > my version seems simpler and easier to maintain down the line
> 
> Okay no problem, I have updated with your version of the code. If possible plz see

@sidd-27 small follow up, I think my PR is ready

**Comment by edgarriba** [comment]
LGTM — @sidd-27 can you merge? Thanks!

**Review by Meggison — APPROVED** [review]

**Comment by sidd-27** [comment]
> Commits must have verified signatures.

this is blocking the merge

**Comment by edgarriba** [comment]
LGTM — @sidd-27 can you merge? Thanks!

**Comment by sidd-27** [comment]
@codeguitarmaster  rebase from main and the tests should pass i think
